README.md
Diablo3::Damage::Calc version 0.12
==================================

INSTALLATION
------------

To install this module type the following:

    git clone git://github.com/ysasaki/p5-diablo3-damage-calc.git
    cd p5-diablo3-damage-calc
    cpanm .
    d3-damage.pl --help

HOW TO USE
----------

d3-damage.pl
------------

_One Hand_

    $ d3-damage.pl -p 1100 -cc 12.5 -cd 97 -b 15 -ias 7.5 -mi 300 -ma 700 -a 1.35
    Base Damage    :  5500.00 ( 6325.00)
    Critical Chance:    12.50
    Critical Damage: 10835.00 (12460.25)
    Avg            :  6166.88 ( 7091.91)
    APS            :     1.45 (1.45, 0.00)
    DPS            :  8949.68 (10292.13)

_Dual Wielding_

    $ d3-damage.pl -p 1200 -cc 38.5 -cd 265 -b 15 -ias 7.5 \
    -mi 400 -ma 800 -a 1.20 \
    -mi 300 -ma 700 -a 1.35
    Base Damage    :  6600.00 ( 7590.00)
    Critical Chance:    38.50
    Critical Damage: 24090.00 (27703.50)
    Avg            : 13333.65 (15333.70)
    APS            :     1.56 (1.47, 1.65)
    DPS            : 20825.49 (23949.32)


d3-damage-compare.pl
--------------------

First, you need to write a configuration file like below.

    $ cat example/compare-config
    -p 1200 -cc 38.5 -cd 265 -b 15 -ias 7.5 -mi 400 -ma 800 -a 1.20 -mi 300 -ma 700 -a 1.35
    -p 1200 -cc 42.5 -cd 265 -b 15 -ias 7.5 -mi 600 -ma 800 -a 1.20 -mi 300 -ma 700 -a 1.35
    -p 1250 -cc 38.5 -cd 305 -b 15 -ias 0   -mi 400 -ma 800 -a 1.20 -mi 350 -ma 750 -a 1.20
    # add support for one-line comments beginning with #
    #-p 1250 -cc 48.5 -cd 165 -b 30 -ias 16  -mi 400 -ma 800 -a 1.20
    -p 1250 -cc 48.5 -cd 165 -b 15 -ias 16  -mi 400 -ma 800 -a 1.53


Next, run `d3-damage-compare.pl -f /path/to/your/config`.

    $ d3-damage-compare.pl -f example/compare-config
    ==============================================================================
    -p 1200 -cc 38.5 -cd 265 -b 15 -ias 7.5 -mi 400 -ma 800 -a 1.20 -mi 300 -ma 700 -a 1.35
    ==============================================================================
    Base Damage    :  6600.00 ( 7590.00)
    Critical Chance:    38.50
    Critical Damage: 24090.00 (27703.50)
    Avg            : 13333.65 (15333.70)
    APS            :     1.56 (1.47, 1.65)
    DPS            : 20825.49 (23949.32)

    ==============================================================================
    -p 1200 -cc 42.5 -cd 265 -b 15 -ias 7.5 -mi 600 -ma 800 -a 1.20 -mi 300 -ma 700 -a 1.35
    ==============================================================================
    Base Damage    :  7200.00 ( 8280.00)     9.09%
    Critical Chance:    42.50               10.39%
    Critical Damage: 26280.00 (30222.00)     9.09%
    Avg            : 15309.00 (17605.35)    14.81%
    APS            :     1.56 (1.47, 1.65)   0.00% (  0.00%,   0.00%)
    DPS            : 23910.74 (27497.36)    14.81%

    ==============================================================================
    -p 1250 -cc 38.5 -cd 305 -b 15 -ias 0   -mi 400 -ma 800 -a 1.20 -mi 350 -ma 750 -a 1.20
    ==============================================================================
    Base Damage    :  7187.50 ( 8265.62)     8.90%
    Critical Chance:    38.50                0.00%
    Critical Damage: 29109.38 (33475.78)    20.84%
    Avg            : 15627.42 (17971.54)    17.20%
    APS            :     1.38 (1.38, 1.38) -11.64% ( -6.12%, -16.55%)
    DPS            : 21565.84 (24800.72)     3.56%

    ==============================================================================
    -p 1250 -cc 48.5 -cd 165 -b 15 -ias 16  -mi 400 -ma 800 -a 1.53
    ==============================================================================
    Base Damage    :  7500.00 ( 8625.00)    13.64%
    Critical Chance:    48.50               25.97%
    Critical Damage: 19875.00 (22856.25)   -17.50%
    Avg            : 13501.88 (15527.16)     1.26%
    APS            :     1.77 (1.77, 0.00)  13.63% ( 20.73%, -100.00%)
    DPS            : 23963.13 (27557.60)    15.07%


ABOUT FORMULA
-------------

See [Calculating Dual Wield Character DPS](http://gaming.stackexchange.com/questions/71589/calculating-dual-wield-character-dps).

TODO
----

 - Support Wizard Source and Witch Doctor Mojo

DEPENDENCIES
------------

This module requires these other modules and libraries:

    Class::Accessor::Lite

COPYRIGHT AND LICENCE
---------------------

Copyright (C) 2012 by Yoshihiro Sasaki

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.2 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Diablo3::Damage::Calc

0.13  Fri Jul 13 02:04 2012
    - fix bug: division by zero

0.12  Thu Jul  5 16:30 2012
    - add support for one-line comments beginning with #

0.11  Thu Jul  5 16:30 2012
    - add --increase_attack_speed option

0.10  Wed Jul  4 18:28 2012
    - add d3-damage-compare.pl
    - fix output format

0.02  Tue Jul  3 15:05 2012
    - support dual wielding
    - moved repository to github from gist

0.01  Mon Jul  2 HH:MM 2012
	- original version



d3-damage.pl
#!perl

use v5.14;
use warnings;
use Pod::Usage;
use Getopt::Long;
use Diablo3::Damage::Calc;
use Diablo3::Damage::Weapon;

my $ret = GetOptions(
    \my %o,                     'min|mi=i@',
    'max|ma=i@',                'primary_stat|p=i',
    'critical_hit_chance|cc=f', 'critical_hit_damage|cd=i',
    'attack_speed|a=f@',        'increase_attack_speed|ias=f',
    'buff|b=f',                 'help|h',
);

if ( $o{help} or !$ret ) {
    pod2usage(1);
}

for (qw(max min attack_speed)) {
    unless ( defined $o{$_}->[0] ) {
        warn "Mandatory parameter $_ is missing\n";
        pod2usage(1);
    }
}

for (qw(primary_stat)) {
    unless ( defined $o{$_} ) {
        warn "Mandatory parameter $_ is missing\n";
        pod2usage(1);
    }
}

$o{critical_hit_chance}   //= 5;
$o{critical_hit_damage}   //= 50;
$o{increase_attack_speed} //= 0;
$o{buff}                  //= 0;

my $calc = Diablo3::Damage::Calc->new(
    main_weapon => Diablo3::Damage::Weapon->new(
        map { $_ => $o{$_}->[0] } qw/min max attack_speed/
    ),
    off_hand => Diablo3::Damage::Weapon->new(
        map { $_ => $o{$_}->[1] } qw/min max attack_speed/
    ),
    map { $_ => $o{$_} }
        qw/primary_stat critical_hit_chance critical_hit_damage
        increase_attack_speed buff/
);

print $calc->as_string, "\n";

=head1 SYNOPSIS

d3-damage.pl

    --min | -mi

    --max | -ma

    --attack_speed | -a

    --primary_stat | -p

    --critical_hit_chance | -cc
        default 5

    --critical_hit_damage | -cd
        default 50

    --increase_attack_speed | -ias
        sum of percentage of non-weapon gears. default 0%.

    --buff | -b
        default 0

    --help | -h

=over 4

=item One Hand

    d3-damage.pl -p 1100 -cc 12.5 -cd 97 -b 15 -ias 7.5 -mi 300 -ma 700 -a 1.35 

=item Dual Wielding

    d3-damage.pl -p 1200 -cc 38.5 -cd 265 -b 15 -ab 7.5 \
    -mi 400 -ma 800 -a 1.20 \
    -mi 300 -ma 700 -a 1.35 

=back

=cut


lib/Diablo3/Damage/Calc.pm
package Diablo3::Damage::Calc;

use v5.14;
use warnings;
use Class::Accessor::Lite (
    new => 1,
    ro  => [
        qw/
            main_weapon off_hand primary_stat
            critical_hit_chance critical_hit_damage
            increase_attack_speed buff
            /
    ],
);
use Carp ();

our $VERSION = '0.13';

sub buffed {
    my $self = shift;
    my $buff = ( 1 + $self->buff / 100 );
    return $buff;
}

sub base_damage {
    my $self = shift;

    unless ( $self->{_base} ) {
        my $avg = $self->main_weapon->avg;
        if ( $self->off_hand->is_equipped ) {
            $avg += $self->off_hand->avg;
            $avg /= 2;
        }
        $self->{_base} = $avg * $self->primary_stat / 100;
    }
    my $base = $self->{_base};
    return $base, $base * $self->buffed;

}

sub critical_damage {
    my $self = shift;

    my ( $base, $base_buffed ) = $self->base_damage;
    my $critical_damage = $base * $self->critical_hit_damage_ratio;

    return $critical_damage, $critical_damage * $self->buffed;
}

sub critical_hit_damage_ratio {
    my $self  = shift;
    my $ratio = 1 + $self->critical_hit_damage / 100;
    return $ratio;
}

sub average_damage {
    my $self = shift;

    my ($base)     = $self->base_damage;
    my ($critical) = $self->critical_damage;
    my $chance     = $self->critical_hit_chance;
    my $avg = ( 1 - $chance / 100 ) * $base + ( $chance / 100 ) * $critical;

    return $avg, $avg * $self->buffed;
}

sub attack_per_second {
    my $self = shift;

    my $is_dual = $self->off_hand->is_equipped ? 1 : 0;

    my $main
        = $is_dual
        ? $self->main_weapon->attack_speed
        * ( 1 + 0.15 + $self->increase_attack_speed / 100 )
        : $self->main_weapon->attack_speed
        * ( 1 + $self->increase_attack_speed / 100 );

    my $off
        = $is_dual
        ? $self->off_hand->attack_speed
        * ( 1 + 0.15 + $self->increase_attack_speed / 100 )
        : 0;

    my $aps
        = $is_dual
        ? ( $main + $off ) / 2
        : $main;

    return $aps, $main, $off;
}

sub as_string {
    my $self = shift;

    my @line;
    my ( $base, $base_buffed ) = $self->base_damage;
    push @line, sprintf "Base Damage    : %8.2f (%8.2f)", $base, $base_buffed;

    my ( $critical_damage, $critical_damage_buffed ) = $self->critical_damage;
    push @line, sprintf "Critical Chance: %8.2f", $self->critical_hit_chance;
    push @line, sprintf "Critical Damage: %8.2f (%8.2f)", $critical_damage,
        $critical_damage_buffed;

    my ( $avg, $avg_buffed ) = $self->average_damage;
    push @line, sprintf "Avg            : %8.2f (%8.2f)", $avg, $avg_buffed;

    my ( $aps, $main_aps, $off_aps ) = $self->attack_per_second;
    push @line, sprintf "APS            : %8.2f (%3.2f, %3.2f)", $aps, $main_aps,
        $off_aps;
    push @line, sprintf "DPS            : %8.2f (%8.2f)", $aps * $avg,
        $aps * $avg_buffed;

    return join "\n", @line;
}

sub as_compared_string {
    my $self      = shift;
    my $base_calc = shift;
    unless ($base_calc) {
        Carp::carp "as_compared_string is required base Calc object.";
        return $self->as_string;
    }

    my @line;

    # Base
    my ( $base, $base_buffed ) = $self->base_damage;
    my ($base_c) = $base_calc->base_damage;
    push @line, sprintf "Base Damage    : %8.2f (%8.2f)   %6.2f%%",
        $base, $base_buffed, _per( $base, $base_c );

    # Critical
    my ( $critical_damage, $critical_damage_buffed ) = $self->critical_damage;
    my ($critical_damage_c) = $base_calc->critical_damage;
    push @line, sprintf "Critical Chance: %8.2f              %6.2f%%",
        $self->critical_hit_chance,
        _per( $self->critical_hit_chance, $base_calc->critical_hit_chance );

    push @line, sprintf "Critical Damage: %8.2f (%8.2f)   %6.2f%%",
        $critical_damage,
        $critical_damage_buffed, _per( $critical_damage, $critical_damage_c );

    # AVG
    my ( $avg, $avg_buffed ) = $self->average_damage;
    my ($avg_c) = $base_calc->average_damage;
    push @line, sprintf "Avg            : %8.2f (%8.2f)   %6.2f%%", $avg,
        $avg_buffed, _per( $avg, $avg_c );

    # APS, DPS
    my ( $aps,   $main_aps,   $off_aps )   = $self->attack_per_second;
    my ( $aps_c, $main_aps_c, $off_aps_c ) = $base_calc->attack_per_second;
    push @line,
        sprintf
        "APS            : %8.2f (%3.2f, %3.2f) %6.2f%% (%6.2f%%, %6.2f%%)",
        $aps, $main_aps, $off_aps,
        _per( $aps, $aps_c ), _per( $main_aps, $main_aps_c ),
        _per( $off_aps, $off_aps_c );

    push @line, sprintf "DPS            : %8.2f (%8.2f)   %6.2f%%", $aps * $avg,
        $aps * $avg_buffed, _per( $aps * $avg, $aps_c * $avg_c );

    return join "\n", @line;
}

sub _per {
    return 0 unless $_[1];
    100 * (($_[0]/$_[1]) - 1)
}


=head1 NAME

Diablo3::Damage::Calc - Perl extention for calculating damages of Diablo3

=head1 SYNOPSIS

See C<d3-damage.pl --help> or C<d3-damage-compare.pl --help>

=head1 AUTHOR

Yoshihiro Sasaki E<lt>ysasaki at cpan.org<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 COPYRIGHT

Copyright 2012 IDAC CO.,LTD. All rights reserved.

=cut

1;


d3-damage-compare.pl
#!perl

use v5.14;
use warnings;
use Pod::Usage;
use Getopt::Long;
use Diablo3::Damage::Calc;
use Diablo3::Damage::Weapon;

my $ret = GetOptions( \my %o, 'file|f=s', 'help|h', );

if ( $o{help} or !$ret ) {
    pod2usage(1);
}

for (qw(file)) {
    unless ( defined $o{$_} ) {
        warn "Mandatory parameter $_ is missing.\n";
        pod2usage(1);
    }
}

open my $fh, '<', $o{file} or die "Cannot open $o{file}: $!";
my @lines = grep !/^#/, <$fh>;
close $fh;

if ( @lines < 2 ) {
    warn "Two ore more settings are required in config file.\n";
    pod2usage(1);
}

my ( $base, @others ) = create_calc_from_config(@lines);

line();
print $base->{_args};
line();
print $base->as_string, "\n\n";
for (@others) {
    line();
    print $_->{_args};
    line();
    print $_->as_compared_string($base), "\n\n";
}

sub line {
    print "=" x 78, "\n";
}

sub create_calc_from_config {
    my @lines = @_;

    my @calc;
    my $i = 1;
    for my $args (@lines) {
        local @ARGV = split /\s+/, $args;
        my $ret = GetOptions(
            \my %o,                     'min|mi=i@',
            'max|ma=i@',                'primary_stat|p=i',
            'critical_hit_chance|cc=f', 'critical_hit_damage|cd=i',
            'attack_speed|a=f@',        'increase_attack_speed|ias=f',
            'buff|b=f',
        );

        if ( !$ret ) {
            die "Something wrong with parameters [$args] at line $i.\n";
        }

        for (qw(max min attack_speed)) {
            unless ( defined $o{$_}->[0] ) {
                die "Mandatory parameter $_ is missing at line $i\n";
            }
        }

        for (qw(primary_stat)) {
            unless ( defined $o{$_} ) {
                die "Mandatory parameter $_ is missing at line $i\n";
            }
        }

        $o{critical_hit_chance}   //= 5;
        $o{critical_hit_damage}   //= 50;
        $o{increase_attack_speed} //= 0;
        $o{buff}                  //= 0;

        my $calc = Diablo3::Damage::Calc->new(
            main_weapon => Diablo3::Damage::Weapon->new(
                map { $_ => $o{$_}->[0] } qw/min max attack_speed/
            ),
            off_hand => Diablo3::Damage::Weapon->new(
                map { $_ => $o{$_}->[1] } qw/min max attack_speed/
            ),
            map { $_ => $o{$_} }
                qw/primary_stat critical_hit_chance critical_hit_damage
                increase_attack_speed buff/
        );

        # FIXME
        $calc->{_args} = $args;

        $i++;
        push @calc, $calc;
    }

    return @calc;
}

=head1 SYNOPSIS

d3-damage-compare.pl

    --file | -f CONFIG

    --help | -h

=head2 CONFIG

Write configurations per line.

    $ cat my-barb-config
    -p 1200 -cc 38.5 -cd 265 -b 15 -ias 7.5 -mi 400 -ma 800 -a 1.20 -mi 300 -ma 700 -a 1.35
    -p 1200 -cc 42.5 -cd 265 -b 15 -ias 7.5 -mi 600 -ma 800 -a 1.20 -mi 300 -ma 700 -a 1.35
    -p 1250 -cc 38.5 -cd 305 -b 15 -ias 0   -mi 400 -ma 800 -a 1.20 -mi 350 -ma 750 -a 1.20
    # add support for one-line comments beginning with #
    #-p 1250 -cc 48.5 -cd 165 -b 30 -ias 16  -mi 400 -ma 800 -a 1.20
    -p 1250 -cc 48.5 -cd 165 -b 15 -ias 16  -mi 400 -ma 800 -a 1.53

=head2 Example

    d3-damage-diff.pl -f my-barb-config
    ==============================================================================
    -p 1200 -cc 38.5 -cd 265 -b 15 -ias 7.5 -mi 400 -ma 800 -a 1.20 -mi 300 -ma 700 -a 1.35
    ==============================================================================
    Base Damage    :  6600.00 ( 7590.00)
    Critical Chance:    38.50
    Critical Damage: 24090.00 (27703.50)
    Avg            : 13333.65 (15333.70)
    APS            :     1.56 (1.47, 1.65)
    DPS            : 20825.49 (23949.32)

    ==============================================================================
    -p 1200 -cc 42.5 -cd 265 -b 15 -ias 7.5 -mi 600 -ma 800 -a 1.20 -mi 300 -ma 700 -a 1.35
    ==============================================================================
    Base Damage    :  7200.00 ( 8280.00)     9.09%
    Critical Chance:    42.50               10.39%
    Critical Damage: 26280.00 (30222.00)     9.09%
    Avg            : 15309.00 (17605.35)    14.81%
    APS            :     1.56 (1.47, 1.65)   0.00% (  0.00%,   0.00%)
    DPS            : 23910.74 (27497.36)    14.81%

    ==============================================================================
    -p 1250 -cc 38.5 -cd 305 -b 15 -ias 0   -mi 400 -ma 800 -a 1.20 -mi 350 -ma 750 -a 1.20
    ==============================================================================
    Base Damage    :  7187.50 ( 8265.62)     8.90%
    Critical Chance:    38.50                0.00%
    Critical Damage: 29109.38 (33475.78)    20.84%
    Avg            : 15627.42 (17971.54)    17.20%
    APS            :     1.38 (1.38, 1.38) -11.64% ( -6.12%, -16.55%)
    DPS            : 21565.84 (24800.72)     3.56%

    ==============================================================================
    -p 1250 -cc 48.5 -cd 165 -b 15 -ias 16  -mi 400 -ma 800 -a 1.53
    ==============================================================================
    Base Damage    :  7500.00 ( 8625.00)    13.64%
    Critical Chance:    48.50               25.97%
    Critical Damage: 19875.00 (22856.25)   -17.50%
    Avg            : 13501.88 (15527.16)     1.26%
    APS            :     1.77 (1.77, 0.00)  13.63% ( 20.73%, -100.00%)
    DPS            : 23963.13 (27557.60)    15.07%
=cut
